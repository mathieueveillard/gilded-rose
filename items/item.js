"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Item = void 0;
var Item = /** @class */ (function () {
    function Item(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
    return Item;
}());
exports.Item = Item;
