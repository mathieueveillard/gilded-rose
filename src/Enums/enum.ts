export enum ItemType {
  AgedBrie = 'Aged Brie',
  BackstagePasses = 'Backstage passes to a TAFKAL80ETC concert',
  Sulfuras = 'Sulfuras, Hand of Ragnaros',
}
  
export enum Quality {
  MIN = 0,
  MAX = 50,
  INCREASE = 1,
  INCREASE_DOUBLE = 2,
  DECREASE = 1,
}
  
export enum SellIn {
  WARNING = 11,
  SOON = 6,
}