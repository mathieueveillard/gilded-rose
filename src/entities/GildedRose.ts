import Item from './Item';

class GildedRose {
  private readonly _items: Array<Item>;

  constructor(items: Array<Item> = []) {
    this._items = items;
  }

  get items(): Array<Item> {
    return this._items;
  }

  public updateQuality(): Array<Item> {
    this._items.forEach(item => {
      console.log(`Processing item: ${item.name} with sellIn: ${item.sellIn} and quality: ${item.quality}`);

      // Reduce sellIn first for all items except Sulfuras
      if (item.name !== "Sulfuras, Hand of Ragnaros") {
        console.log('Item is not Sulfuras, reducing sellIn by 1');
        item.sellIn--;
      }

      switch (item.name) {
        case "Aged Brie":
          console.log('Item is Aged Brie');
          this.updateAgedBrie(item);
          break;
        case "Backstage passes to a TAFKAL80ETC concert":
          console.log('Item is Backstage passes');
          this.updateBackstagePasses(item);
          break;
        case "Sulfuras, Hand of Ragnaros":
          console.log('Item is Sulfuras, no change required');
          break;
        case "Conjured":
          console.log('Item is Conjured');
          this.updateConjuredItem(item);
          break;
        default:
          console.log('Item is a regular item');
          this.updateNormalItem(item);
          break;
      }

      // Ensure quality remains within valid bounds
      if (item.quality < 0) item.quality = 0;
      if (item.quality > 50 && item.name !== "Sulfuras, Hand of Ragnaros") item.quality = 50;
    });

    return this._items;
  }

  private updateAgedBrie(item: Item) {
    if (item.quality < 50) {
      console.log('Increasing quality by 1 for Aged Brie');
      item.quality++;
    }
    if (item.sellIn < 0 && item.quality < 50) {
      console.log('SellIn is negative, increasing quality by 1 again for Aged Brie');
      item.quality++;
    }
  }

  private updateBackstagePasses(item: Item) {
    if (item.quality < 50) {
      console.log('Increasing quality by 1 for Backstage passes');
      item.quality++;
      if (item.sellIn < 11 && item.quality < 50) {
        console.log('SellIn is less than 11, increasing quality by 1 again');
        item.quality++;
      }
      if (item.sellIn < 6 && item.quality < 50) {
        console.log('SellIn is less than 6, increasing quality by 1 once more');
        item.quality++;
      }
    }
    if (item.sellIn < 0) {
      console.log('Setting Backstage passes quality to 0 after concert');
      item.quality = 0;
    }
  }

  private updateNormalItem(item: Item) {
    if (item.quality > 0) {
      console.log('Reducing quality by 1 for regular item');
      item.quality--;
    }
    if (item.sellIn < 0 && item.quality > 0) {
      console.log('SellIn is negative, reducing quality by 1 again for regular item');
      item.quality--;
    }
  }

  private updateConjuredItem(item: Item) {
    if (item.quality > 0) {
      console.log('Reducing quality by 2 for Conjured item');
      item.quality -= 2;
    }
    if (item.sellIn < 0 && item.quality > 0) {
      console.log('SellIn is negative, reducing quality by 2 again for Conjured item');
      item.quality -= 2;
    }
    if (item.quality < 0) {
      console.log('Quality is negative, setting it to 0');
      item.quality = 0;  // Ensure quality is never negative
    }
  }
}

export default GildedRose;
