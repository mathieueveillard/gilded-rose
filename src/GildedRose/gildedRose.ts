import { Item } from "../Items/item";
import { ItemType, Quality, SellIn } from "../Enums/enum";

export class GildedRose {
  items: Array<Item>;

  constructor(items: Array<Item> = []) {
    this.items = items;
  }

  updateQuality() {
    for (const item of this.items) {
      console.log(`Before updateQuality: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

      switch (item.name) {
        case ItemType.AgedBrie:
          this.updateAgedBrie(item);
          break;
        case ItemType.BackstagePasses:
          this.updateBackstagePasses(item);
          break;
        case ItemType.Sulfuras:
          break;
        default:
          this.updateStandardItem(item);
          break;
      }

      console.log(`After updateQuality: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
    }

    return this.items;
  }

  private updateAgedBrie(item: Item) {
    console.log(`Before updateAgedBrie: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

    item.sellIn -= 1;
    item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE);

    console.log(`After updateAgedBrie: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
  }

  private updateBackstagePasses(item: Item) {
    console.log(`Before updateBackstagePasses: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

    item.sellIn -= 1;

    if (item.sellIn < 0) {
      item.quality = Quality.MIN;
    } else if (item.sellIn <= SellIn.WARNING) {
      item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE_DOUBLE);
    } else if (item.sellIn <= SellIn.SOON) {
      item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE);
    }

    console.log(`After updateBackstagePasses: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
  }

  private updateStandardItem(item: Item) {
    console.log(`Before updateStandardItem: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
    
    item.sellIn -= 1;
    
    // Diminue la qualité deux fois plus rapidement si la date de vente est passée
    if (item.sellIn < 0) {
        item.quality = Math.max(Quality.MIN, item.quality - Quality.DECREASE * 2);
    } else {
        item.quality = Math.max(Quality.MIN, item.quality - Quality.DECREASE);
    }
    
    // Assurez-vous que la qualité ne dépasse pas 50
    if (item.quality > Quality.MAX) {
        item.quality = Quality.MAX;
    }

    console.log(`After updateStandardItem: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
}
}
