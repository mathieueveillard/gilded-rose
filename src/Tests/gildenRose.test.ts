import { GildedRose } from "../GildedRose/gildedRose";
import { Item } from "../Items/item";
import { ItemType } from "../Enums/enum";


// si un article diminue deux fois plus rapidement une fois la date de vente passée.
test("Once the sell by date has passed, Quality degrades twice as fast", () => {
    const items = [new Item("Standard Item", 0, 20)];
    const gildedRose = new GildedRose(items);
    gildedRose.updateQuality();
    
    expect(items).toMatchSnapshot();
  });
  
  // si la qualité d'un article ne devient jamais négative.
  test("The Quality of an item is never negative", () => {
    const items = [new Item("Standard Item", 10, 0)];
    const gildedRose = new GildedRose(items);
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();

  });
  
  //si la qualité de l'article "Aged Brie" augmente à mesure qu'il vieillit.
  test("Aged Brie actually increases in Quality the older it gets", () => {
    const items = [new Item("Aged Brie", 10, 20)];
    const gildedRose = new GildedRose(items);
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();

  });

  // si la qualité d'un article ne dépasse jamais la valeur de 50.
  test("The Quality of an item is never more than 50", () => {
    const items = [new Item("Standard Item", 10, 60)];
    const gildedRose = new GildedRose(items);
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
 
  });
  