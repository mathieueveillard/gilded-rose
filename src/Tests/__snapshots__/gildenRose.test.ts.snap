// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aged Brie actually increases in Quality the older it gets 1`] = `
[
  Item {
    "name": "Aged Brie",
    "quality": 21,
    "sellIn": 9,
  },
]
`;

exports[`Once the sell by date has passed, Quality degrades twice as fast 1`] = `
[
  Item {
    "name": "Standard Item",
    "quality": 18,
    "sellIn": -1,
  },
]
`;

exports[`The Quality of an item is never more than 50 1`] = `
[
  Item {
    "name": "Standard Item",
    "quality": 59,
    "sellIn": 9,
  },
]
`;

exports[`The Quality of an item is never negative 1`] = `
[
  Item {
    "name": "Standard Item",
    "quality": 0,
    "sellIn": 9,
  },
]
`;
